src/Popover.jsx

```diff
@@ -0,0 +1,206 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import cx from 'classnames';
+import bindAll from './utils/bindAll';
+
+/**
+ * @module Popover
+ */
+class Popover extends React.Component {
+	constructor(props) {
+		super(props);
+
+		bindAll(this,
+			'updateFocusBy',
+			'openMenu',
+			'closeMenu',
+			'onClick',
+			'onKeyUp',
+			'onKeyDown',
+			'onClick',
+			'onBlur'
+		);
+
+		this.state = {
+			isActive: false,
+			selectedIndex: 0
+		};
+	}
+
+	updateFocusBy(delta) {
+		const targetIndex = this.state.selectedIndex + delta;
+		const optionsLength = this.props.options.length;
+
+		if (targetIndex >= 0 && targetIndex <= optionsLength) {
+			this.setState({ selectedIndex: targetIndex });
+		}
+	}
+
+	openMenu() {
+		this.setState({ isActive: true });
+	}
+
+	closeMenu() {
+		this.setState({ isActive: false });
+	}
+
+	onBlur() {
+		// On blur, browsers always focus `<body>` before moving focus
+		// to the next actual focused element.
+		//
+		// This zero-length timeout ensures the browser will return the
+		// actual focused element instead of `<body>`
+		window.setTimeout(() => {
+			const focusedOptionClass = document.activeElement.parentNode.getAttribute('class');
+
+			// don't close the popover if we're moving focus to an option
+			if (focusedOptionClass && focusedOptionClass.indexOf('popover-menu-option') > -1) {
+				return;
+			}
+
+			this.closeMenu();
+		}, 0);
+	}
+
+	onClick(e) {
+		console.warn('click!');
+		this.openMenu();
+	}
+
+	onKeyDown(e) {
+		switch(e.key) {
+		case 'Enter':
+			if (this.state.isActive
+				&& this.selectedItemEl
+				&& this.selectedItemEl.props.onClick
+			) {
+				this.selectedItemEl.props.onClick(e);
+				break;
+			} else {
+				this.openMenu();
+				break;
+			}
+		case 'Escape':
+			this.closeMenu();
+			break;
+		}
+	}
+
+	onKeyUp(e) {
+		switch(e.key) {
+		case 'ArrowDown':
+			this.updateFocusBy(1);
+			break;
+		case 'ArrowUp':
+			this.updateFocusBy(-1);
+			break;
+		}
+	}
+
+	componentDidUpdate() {
+		if (this.selectedItemEl) {
+			ReactDOM.findDOMNode(this.selectedItemEl).focus();
+		}
+	}
+
+	render() {
+		const isActive = this.state.isActive,
+			{
+				trigger,
+				options,
+				className,
+				...other
+			} = this.props,
+			{
+				onClick,
+				onKeyUp,
+				onKeyDown,
+				onBlur
+			} = this;
```

Comment:
Nitpick, but destructuring instance properties into new `const`s feels like an over-abstraction that makes usage more ambiguous - leaving them as `this.x` makes it clear that each method is an instance method rather than a standalone function defined somewhere, with the associated expectation of context binding.